presto-memory-context/src/main/java/com/facebook/presto/memory/AggregatedMemoryContext.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.memory;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class AggregatedMemoryContext
+{
+    @Nullable
+    @GuardedBy("this")
+    private final AggregatedMemoryContext parentMemoryContext;
+    @Nullable
+    @GuardedBy("this")
+    private MemoryNotificationListener notificationListener;
+    @GuardedBy("this")
+    private long usedBytes;
+    @GuardedBy("this")
+    private boolean closed;
+
+    public AggregatedMemoryContext()
+    {
+        this.parentMemoryContext = null;
+    }
+
+    private AggregatedMemoryContext(AggregatedMemoryContext parentMemoryContext)
+    {
+        this.parentMemoryContext = requireNonNull(parentMemoryContext, "parentMemoryContext is null");
+    }
+
+    public AggregatedMemoryContext newAggregatedMemoryContext()
+    {
+        return new AggregatedMemoryContext(this);
+    }
+
+    public LocalMemoryContext newLocalMemoryContext()
+    {
+        return new LocalMemoryContext(this);
+    }
+
+    public synchronized long getBytes()
+    {
+        return usedBytes;
+    }
+
+    protected synchronized void updateBytes(long bytes)
+    {
+        if (parentMemoryContext != null) {
+            parentMemoryContext.updateBytes(bytes);
+        }
+        long oldLocalUsedBytes = this.usedBytes;
+        usedBytes += bytes;
+        if (notificationListener != null) {
+            notificationListener.memoryUsageUpdated(oldLocalUsedBytes, usedBytes);
```

Comment:
If you need the listener invoked inside the object monitor, add a comment that explains why.

My understanding is that you need the order guarantee provided by the object monitor. I'm not sure whether I'm correct.